### ISSUE TITLE
[change] ModemManager device option needs to update to ctl_device

### ISSUE THREAD

**Issue by ghost** [openwisp/openwrt-openwisp-monitoring#70]
Hey,

Just letting you know, we are changing this from `device` to `ctl_device` upstream in the OpenWrt ModemManager protocol handler because of compatibility issues with DSA.  I will link the PR here when it is in.

https://github.com/openwisp/openwrt-openwisp-monitoring/blob/261239e7f6b0ff86876aebcb11292722e59c0c78/openwrt-openwisp-monitoring/files/lib/openwisp_monitoring/interfaces.lua#L21

**Comment by nickberry17** [comment]
How's this look?  It looks like the `uci_cursor.get` returns `nil` if that entry doesn't exist, right?

```

        local compat_device = uci_cursor.get('network', interface['interface'], 'device')
        local device = uci_cursor.get('network', interface['interface'], 'ctl_device')
        local modem = (compat_device or device)

        local info = {}

        local general = io.popen('mmcli --output-json -m '..modem):read("*a")
```

**Comment by nemesisdesign** [comment]
@nickberry17 yes, it should, I haven't tested the following modified version yet but if it works is a bit shorter but still readable:

```lua
local device = uci_cursor.get('network', interface['interface'], 'ctl_device') or \
               uci_cursor.get('network', interface['interface'], 'device')
local info = {}
local general = io.popen('mmcli --output-json -m '..device):read("*a")
```

The second lineand the or won't be executed in the future and can be removed at some point in the future.

**Comment by nickberry17** [comment]
Is it an inclusive `or`? Do we need to check if both are set, so we don't get `device` set to `/sys/devices/platform/soc/8af8800.usb3/8a00000.dwc3/xhci-hcd.0.auto/usb2/2-1/sys/devices/platform/soc/8af8800.usb3/8a00000.dwc3/xhci-hcd.0.auto/usb2/2-1`

**Comment by nemesisdesign** [comment]
> Is it an inclusive `or`? Do we need to check if both are set, so we don't get `device` set to `/sys/devices/platform/soc/8af8800.usb3/8a00000.dwc3/xhci-hcd.0.auto/usb2/2-1/sys/devices/platform/soc/8af8800.usb3/8a00000.dwc3/xhci-hcd.0.auto/usb2/2-1` In a sysupgrade scenario where settings are kept, both `device` and `ctl_device` could be set.

The one with precedence should come first, the or is executed only if the first value is `nil`.

**Comment by nickberry17** [comment]
> @nickberry17 yes, it should, I haven't tested the following modified version yet but if it works is a bit shorter but still readable:
> 
> ```lua
> local device = uci_cursor.get('network', interface['interface'], 'ctl_device') or \
>                uci_cursor.get('network', interface['interface'], 'device')
> local info = {}
> local general = io.popen('mmcli --output-json -m '..device):read("*a")
> ```
> 
> The second lineand the or won't be executed in the future and can be removed at some point in the future.

Didn't like the `/` 

```
uci: Entry not found
lua: /usr/sbin/netjson-monitoring:326: unexpected symbol near '\'
```
Putting them on the same line:

```
uci: Entry not found
lua: /usr/sbin/netjson-monitoring:330: attempt to concatenate global 'modem' (a nil value)
stack traceback:
	/usr/sbin/netjson-monitoring:330: in function 'specialized_info'
	/usr/sbin/netjson-monitoring:392: in function 'get_interface_info'
	/usr/sbin/netjson-monitoring:578: in main chunk
	[C]: ?
```

The longhand way works though

**Comment by devkapilbansal** [comment]
> Putting them on the same line:
> 
> ```
> uci: Entry not found
> lua: /usr/sbin/netjson-monitoring:330: attempt to concatenate global 'modem' (a nil value)
> stack traceback:
> 	/usr/sbin/netjson-monitoring:330: in function 'specialized_info'
> 	/usr/sbin/netjson-monitoring:392: in function 'get_interface_info'
> 	/usr/sbin/netjson-monitoring:578: in main chunk
> 	[C]: ?
> ```
> 

What's the modem here? I think it is renamed to device in other prior instances expect one giving error

**Comment by shwetd19** [comment]
Hey @nemesifier Here's how I think we can solve it : 

1. Locate the Lua script where the `device` option is being used.
2. Replace `device` with `ctl_device` and add backward compatibility:
   ```lua
   local device = uci_cursor.get('network', interface['interface'], 'ctl_device') or
                  uci_cursor.get('network', interface['interface'], 'device')
   ```
3. after testing the changes to ensure the ModemManager functionality works as expected.

**Comment by nemesifier** [comment]
@shwetd19 sounds good but we'll need to test this with OpenWrt 24.10 on a router which has modem-manager, It may take some time before we can test this effectively, but it seems like an easy one so you can open the PR and then we can seek feedback.

**Comment by PRoIISHAAN** [comment]
@openwisp-companion assign
